### PR TITLE
ascanrules: Using timing utils in MYSQL

### PR DESCRIPTION
**Pull request by itsdivyanshjain**
## Overview
Briefly describe the purpose, goals, and changes or improvements made in this pull request.

## Related Issues
Specify any related issues or pull requests by linking to them.

## Checklist
- [ ] Update help
- [X] Update changelog
- [X] Run `./gradlew spotlessApply` for code formatting
- [X] Write tests
- [ ] Check code coverage
- [ ] Sign-off commits
- [ ] Squash commits
- [X] Use a descriptive title

For more details, please refer to the [developer rules and guidelines](https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/).

### CODE REVIEW

File: addOns/ascanrules/CHANGELOG.md [hunk 1]
```diff
@@ -9,6 +9,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 - Improved the following scan rules by using time-based linear regression tests:
   - SQL Injection - Hypersonic SQL
   - SQL Injection - MsSQL
+  - SQL Injection - MySQL
```
Comment (kingthorin) [inline addOns/ascanrules/CHANGELOG.md:12]: Isn't that what 11 says?

**Comment by kingthorin** [comment]
Ugh I'm mixing up MsSQL and MySQL, will look in the morning.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java [hunk 2]
```diff
@@ -269,187 +289,96 @@ public void init() {
...
-            }
+        LOGGER.debug(
+                "Scanning URL [{}] [{}], field [{}] with value [{}] for SQL Injection",
+                getBaseMsg().getRequestHeader().getMethod(),
+                getBaseMsg().getRequestHeader().getURI(),
+                paramName,
+                originalParamValue);
+
+        AtomicReference<HttpMessage> message = new AtomicReference<>();
+        AtomicReference<String> attack = new AtomicReference<>();
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java:300]: Move inside the loop.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java [hunk 3]
```diff
@@ -269,187 +289,96 @@ public void init() {
...
+            String sleepPayload = it.next();
+            TimingUtils.RequestSender requestSender =
+                    x -> {
+                        HttpMessage msg = getNewMsg();
+                        message.compareAndSet(null, msg);
+
+                        String finalPayload =
+                                sleepPayload
+                                        .replace(ORIG_VALUE_TOKEN, originalParamValue)
+                                        // Time in milliseconds for the SQL function.
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java:312]: Remove, not the case here.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java [hunk 4]
```diff
@@ -269,187 +289,96 @@ public void init() {
...
+                        return msg.getTimeElapsedMillis() / 1000.0;
+                    };
             // end of timing baseline check
 
-            int countTimeBasedRequests = 0;
+            String attackPayload =
+                    sleepPayload
+                            .replace(ORIG_VALUE_TOKEN, originalParamValue)
+                            // Time in milliseconds for the SQL function.
+                            .replace(SLEEP_TOKEN, String.valueOf(timeSleepSeconds));
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java:328]: Can be removed with the `AtomicReference`s inside the loop.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java [hunk 5]
```diff
@@ -269,187 +289,89 @@ public void init() {
...
+                                        .replace(SLEEP_TOKEN, String.valueOf(x));
+
+                        setParameter(msg, paramName, finalPayload);
+                        LOGGER.debug("Testing [{}] = [{}]", paramName, finalPayload);
+                        attack.compareAndSet(null, finalPayload);
+
+                        sendAndReceive(msg, false);
+                        return msg.getTimeElapsedMillis() / 1000.0;
+                    };
             // end of timing baseline check
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java:321]:
```suggestion
```

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java [hunk 6]
```diff
@@ -269,187 +289,89 @@ public void init() {
...
+                LOGGER.debug(
+                        "Caught {} {} when accessing: {}.\n The target may have replied with a poorly formed redirect due to our input.",
+                        ex.getClass().getName(),
+                        ex.getMessage(),
+                        message.get().getRequestHeader().getURI());
+            } catch (IOException ex) {
+                LOGGER.debug(
+                        "Check failed for parameter [{}] and payload [{}] due to an I/O error",
+                        paramName,
+                        "attack.get()",
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java:367]:
```suggestion
                        attack.get(),
```

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java [hunk 7]
```diff
@@ -269,187 +289,88 @@ public void init() {
...
+            String sleepPayload = it.next();
+            TimingUtils.RequestSender requestSender =
+                    x -> {
+                        HttpMessage msg = getNewMsg();
+                        message.compareAndSet(null, msg);
+
+                        String finalPayload =
+                                sleepPayload
+                                        .replace(ORIG_VALUE_TOKEN, originalParamValue)
+                                        .replace(SLEEP_TOKEN, String.valueOf(x));
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRule.java:312]:
```suggestion
                                        .replace(SLEEP_TOKEN, Integer.toString((int) x));
```

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRuleUnitTest.java [hunk 8]
```diff
@@ -90,13 +97,13 @@ protected Response serve(IHTTPSession session) {
...
         this.rule.init(msg, this.parent);
-        this.rule.setSleepInSeconds(1);
+        this.rule.setSleepInSeconds(2);
 
         this.rule.scan();
 
         assertThat(alertsRaised.size(), equalTo(1));
         assertThat(alertsRaised.get(0).getParam(), equalTo("name"));
-        assertThat(alertsRaised.get(0).getAttack(), equalTo("test' / sleep(1) / '"));
+        assertThat(alertsRaised.get(0).getAttack(), equalTo("test' / sleep(2.0) / '"));
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRuleUnitTest.java:106]:
```suggestion
        assertThat(alertsRaised.get(0).getAttack(), equalTo("test' / sleep(2) / '"));
```

File: addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRuleUnitTest.java [hunk 9]
```diff
@@ -66,22 +68,27 @@ void shouldNotTargetNonMySQLTechs() throws Exception {
     @Test
     void shouldAlertIfSleepTimesGetLonger() throws Exception {
         String test = "/shouldReportSqlTimingIssue/";
+        Pattern sleepPattern = Pattern.compile("'\\s+/\\s+sleep\\((\\d+)\\.\\d+\\)\\s+/\\s+'");
```
Comment (thc202) [inline addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionMySqlScanRuleUnitTest.java:71]:
```suggestion
        Pattern sleepPattern = Pattern.compile("'\\s+/\\s+sleep\\((\\d+)\\)\\s+/\\s+'");
```

**Review by kingthorin — APPROVED** [review]

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Pull request merged**